### PR TITLE
Make data_start/end weak symbols

### PR DESCRIPTION
**Pull request by jdinan**
Fix LD_PRELOAD error encountered by tools folks.

### CODE REVIEW

File: src/init.c [hunk 1]
```diff
@@ -209,8 +213,12 @@ shmem_internal_init(int tl_requested, int *tl_provided)
     shmem_internal_data_base = (void*) get_etext();
     shmem_internal_data_length = get_end() - get_etext();
 #else
-    shmem_internal_data_base = &data_start;
-    shmem_internal_data_length = (unsigned long) &end  - (unsigned long) &data_start;
+    if (&__data_start == 0 || &_end == 0)
```
Comment (jdinan) [inline src/init.c]: Should probably cast 0 to (int*).

Also, need a comment to record that if the weak symbol isn't bound, its address will be 0.

**Comment by jdinan** [comment]
@davidozog Ready for review

File: configure.ac [hunk 2]
```diff
@@ -594,20 +594,26 @@ AC_TRY_RUN([
...
     base = (void*) get_etext();
     length = get_end() - get_etext();
 #else
-    base = &data_start;
-    length = (unsigned long) &end  - (unsigned long) &data_start;
+    if (&__data_start == (int*) 0 || &_end == (int*) 0) return 1;
+    base = (void*) &__data_start;
+    length = (long) ((char*) &_end - (char*) &__data_start);
 #endif
+    if (!(base != (void*)0 && length > 0)) return 2;
```
Comment (davidozog) [inline configure.ac]: The `length>0` seems like a good check.  Do we also want it in `src/init.c`?  Or is it checked further along?

File: configure.ac [hunk 3]
```diff
@@ -594,20 +594,26 @@ AC_TRY_RUN([
...
+    long length;
 #ifdef __APPLE__
     base = (void*) get_etext();
     length = get_end() - get_etext();
 #else
-    base = &data_start;
-    length = (unsigned long) &end  - (unsigned long) &data_start;
+    if (&__data_start == (int*) 0 || &_end == (int*) 0) return 1;
+    base = (void*) &__data_start;
+    length = (long) ((char*) &_end - (char*) &__data_start);
```
Comment (davidozog) [inline configure.ac]: The `(char*)` cast is slightly perplexing to me, but it looks like it works out...  Do we need it?  E.g., would `(long)&end - (long)&__data_start` not work?

File: src/init.c [hunk 4]
```diff
@@ -209,8 +213,26 @@ shmem_internal_init(int tl_requested, int *tl_provided)
...
+     * assign them an lvalue of 0.  Here, we check that the binary that
+     * initializes OpenSHMEM has exposed these symbols, enabling the library to
+     * locate its symmetric data segment. */
+
+    if (&__data_start == (int*) 0 || &_end == (int*) 0)
+        RETURN_ERROR_MSG("Unable to locate symmetric data segment (%p, %p)\n",
+                         (void*) &__data_start, (void*) &_end);
+
+    shmem_internal_data_base = (void*) &__data_start;
+    shmem_internal_data_length = (long) ((char*) &_end - (char*) &__data_start);
```
Comment (davidozog) [inline src/init.c]: Same question as above regarding the `(char*)` cast.

File: configure.ac (hunk 3 above)
Comment (jdinan) [reply configure.ac]: Casting a pointer to an integer is bad (unless it's intptr_t); it's better to do the arithmetic on the pointer type.  If we do the arithmetic on int*, we need to multiply the result by sizeof(int).  Casting to char* tells us directly how many bytes, which is the number we need.

File: configure.ac (hunk 2 above)
Comment (jdinan) [reply configure.ac]: A zero-length data segment is technically allowed (although I'm not sure we've ever tested that this works).  The configury test puts something in the data segment, so we expect the size to be nonzero.

File: configure.ac (hunk 3 above)
Comment (davidozog) [reply configure.ac]: Ah yes, agreed.

**Pull request merged**